core-primitives/extrinsics-factory/src/lib.rs

```diff
@@ -0,0 +1,162 @@
+	NonceCache: MutateNonce,
+{
+	genesis_hash: H256,
+	signer: Signer,
+	nonce_cache: Arc<NonceCache>,
+}
+
+impl<Signer, NonceCache> ExtrinsicsFactory<Signer, NonceCache>
+where
+	Signer: Pair<Public = sp_core::ed25519::Public>,
+	Signer::Signature: Into<MultiSignature>,
+	NonceCache: MutateNonce,
+{
+	pub fn new(genesis_hash: H256, signer: Signer, nonce_cache: Arc<NonceCache>) -> Self {
+		ExtrinsicsFactory { genesis_hash, signer, nonce_cache }
+	}
+}
+
+impl<Signer, NonceCache> CreateExtrinsics for ExtrinsicsFactory<Signer, NonceCache>
+where
+	Signer: Pair<Public = sp_core::ed25519::Public>,
+	Signer::Signature: Into<MultiSignature>,
+	NonceCache: MutateNonce,
+{
+	fn create_extrinsics(&self, calls: &[OpaqueCall]) -> Result<Vec<OpaqueExtrinsic>> {
+		let mut nonce_lock = self.nonce_cache.load_for_mutation()?;
+		let mut nonce_value = nonce_lock.0;
+
+		let extrinsics_buffer: Vec<OpaqueExtrinsic> = calls
+			.iter()
+			.map(|call| {
+				let xt = compose_extrinsic_offline!(
+					self.signer.clone(),
+					call,
+					nonce_value,
+					Era::Immortal,
+					self.genesis_hash,
+					self.genesis_hash,
+					RUNTIME_SPEC_VERSION,
+					RUNTIME_TRANSACTION_VERSION
+				)
+				.encode();
+				nonce_value += 1;
+				xt
+			})
+			.map(|xt| {
+				OpaqueExtrinsic::from_bytes(&xt)
+					.expect("A previously encoded extrinsic has valid codec; qed.")
+			})
+			.collect();
+
+		*nonce_lock = Nonce(nonce_value);
+
+		Ok(extrinsics_buffer)
+	}
+}
+
+#[cfg(test)]
+pub mod tests {
+
+	use super::*;
+	use codec::Decode;
+	use itp_nonce_cache::{Nonce, NonceCache, NonceValue};
+	use sp_core::ed25519;
+	use substrate_api_client::extrinsic::xt_primitives::UncheckedExtrinsicV4;
+
+	#[test]
+	pub fn creating_xts_increases_nonce_for_each_xt() {
+		let nonce_cache = Arc::new(NonceCache::default());
+		let extrinsics_factory =
+			ExtrinsicsFactory::new(test_genesis_hash(), test_account(), nonce_cache.clone());
+
+		let opaque_calls = [OpaqueCall(vec![3u8; 42]), OpaqueCall(vec![12u8, 78])];
+		let xts = extrinsics_factory.create_extrinsics(&opaque_calls).unwrap();
+
+		assert_eq!(opaque_calls.len(), xts.len());
+		assert_eq!(nonce_cache.get_nonce().unwrap(), Nonce(opaque_calls.len() as NonceValue));
+	}
+
+	// #[test]
+	// pub fn xts_have_increasing_nonce() {
+	// 	let nonce_cache = Arc::new(NonceCache::default());
+	// 	nonce_cache.set_nonce(Nonce(34)).unwrap();
+	// 	let extrinsics_factory =
+	// 		ExtrinsicsFactory::new(test_genesis_hash(), test_account(), nonce_cache);
+	//
+	// 	let opaque_calls =
+	// 		[OpaqueCall(vec![3u8; 42]), OpaqueCall(vec![12u8, 78]), OpaqueCall(vec![15u8, 12])];
+	// 	let xts: Vec<UncheckedExtrinsicV4<OpaqueCall>> = extrinsics_factory
+	// 		.create_extrinsics(&opaque_calls)
+	// 		.unwrap()
+	// 		.iter()
+	// 		.map(|mut x| UncheckedExtrinsicV4::<OpaqueCall>::decode(&mut x))
+	// 		.collect();
+	//
+	// 	assert_eq!(xts.len(), opaque_calls.len());
+	// 	assert_eq!(xts[0].signature.unwrap().2 .2, 34u128);
+	// }
```

Comment:
I wanted to write a test where I create extrinsics and then check if the nonce in the resulting extrinsics is correct. Unfortunately, the nonce in the extrinsic is in a private member that I cannot access here 😞

Reply:
If you want to be hardcore, you could check the relevant bits in the encoded extrinsics. :D.

Anyhow, we can make the signature public in the api-client. I don't see anything against that.